File: src/util/Random.java
package util;

import annotation.Draft;
import annotation.Required;
import annotation.Intention;
import nio.ForeignMemory;
import oop.TypeRegister;
import objects.Probable;
import objects.ProbableObjects;

import nio.StringLookup;
/**
 * Off-Heap Pseudo-Random Number Generator (PRNG) Subsystem.
 * Provides state-updating PRNG operations.
 */
@Draft
@Intention("Zero-allocation off-heap PRNG using mixed chaotic hash finalizers (Murmur3) on mutable sequence counter state.")
public final class Random {

    @Required
    public static final int CLASS_ID = TypeRegister.ID_RANDOM;

    // 64-bit golden ratio fractional constant (Weyl sequence step) for balanced/even dispersion
    private static final long GOLDEN_RATIO_64 = 0x9e3779b97f4a7c15L;

    private static final long SYSTEM_RNG = allocate(System.nanoTime());

    public static long getSystemRng() {
        return SYSTEM_RNG;
    }

    private Random() {}

    public static int classId() {
        return CLASS_ID;
    }

    /**
     * Allocates a new off-heap Random state block.
     * Block layout:
     *   [userPtr - 8]: 32-bit packed Type ID
     *   [userPtr - 4]: 32-bit length (1)
     *   [userPtr + 0]: 64-bit seed / state
     *   [userPtr + 8]: 32-bit current integer counter
     *   [userPtr + 12]: 32-bit padding
     *
     * @param seed the initial 64-bit seed
     * @return the raw memory address pointer of the user payload state
     */
    @Draft
    public static long allocate(long seed) {
        long block = ForeignMemory.allocateNative(24);
        long userPtr = block + 8L;

        // Write class metadata headers
        ForeignMemory.setInt(block, TypeRegister.RANDOM_SINGLETON);
        ForeignMemory.setInt(block + 4L, 1);

        // Write initial state values
        ForeignMemory.setLong(userPtr, seed);
        ForeignMemory.setInt(userPtr + 8L, 0);
        ForeignMemory.setInt(userPtr + 12L, 0);

        return userPtr;
    }

    /**
     * Frees the allocated native memory block for the given Random state pointer.
     *
     * @param ptr the raw memory address pointer returned by allocate
     */
    @Draft
    public static void free(long ptr) {
        ForeignMemory.freeNative(ptr - 8L);
    }

    // ==========================================
    // STATE-BASED MUTATING GENERATION PIPELINE
    // ==========================================

    @Draft
    public static long nextLong(long ptr) {
        if (ptr == 0L) throw new NullPointerException(StringLookup.getJavaString(27));
        
        long seed = ForeignMemory.getLong(ptr);
        int currentInteger = ForeignMemory.getInt(ptr + 8L);

        // Advance seed/state using chaotic Murmur3 mix and sequence index
        long mixed = seed ^ ptr ^ ((long) currentInteger * GOLDEN_RATIO_64);
        long result = util.Hash.murmur3Mix64(mixed);

        // Commit updated state back to off-heap memory
        ForeignMemory.setLong(ptr, result);
        ForeignMemory.setInt(ptr + 8L, currentInteger + 1);

        return result;
    }

    @Draft
    public static int nextInt(long ptr) {
        return (int) nextLong(ptr);
    }

    @Draft
    public static float nextFloat(long ptr) {
        return (nextLong(ptr) & 0xFFFFFF) / 16777216.0f;
    }

    @Draft
    public static double nextDouble(long ptr) {
        return (nextLong(ptr) & 0x1FFFFFFFFFFFFFL) / 9007199254740992.0;
    }

    @Draft
    public static float nextNDCFloat(long ptr) {
        return ((nextLong(ptr) & 0xFFFFFF) / 8388608.0f) - 1.0f;
    }

    @Draft
    public static char nextChar(long ptr) {
        long val = nextLong(ptr);
        return (char) (32 + (Math.abs(val) % 95)); // Printable ASCII: space (32) to tilde (126)
    }

    @Draft
    public static boolean getWeight(long ptr, int weight, int total) {
        if (ptr == 0L) throw new NullPointerException(StringLookup.getJavaString(27));
        if (total <= 0) return false;
        if (weight >= total) return true;
        if (weight <= 0) return false;
        long val = nextLong(ptr) & 0x7FFFFFFFFFFFFFFFL;
        return (val % total) < weight;
    }

    @Draft
    public static long sample(long ptr, long probablePtr) {
        if (probablePtr == 0L) throw new NullPointerException(StringLookup.getJavaString(27));
        int weight = Probable.getWeight(probablePtr);
        int total = Probable.getTotal(probablePtr);
        if (getWeight(ptr, weight, total)) {
            return Probable.getObject(probablePtr);
        }
        return 0L;
    }

    @Draft
    public static long probablePool(long ptr, long probableObjectsPtr) {
        if (probableObjectsPtr == 0L) throw new NullPointerException(StringLookup.getJavaString(27));
        int count = ProbableObjects.size(probableObjectsPtr);
        if (count == 0) return 0L;

        int totalWeight = ProbableObjects.getTotalWeight(probableObjectsPtr);
        if (totalWeight <= 0) {
            long val = nextLong(ptr) & 0x7FFFFFFFFFFFFFFFL;
            int idx = (int) (val % count);
            long slotBase = probableObjectsPtr + 8L + (idx * 16L);
            return ForeignMemory.getLong(slotBase);
        }

        long val = nextLong(ptr) & 0x7FFFFFFFFFFFFFFFL;
        int r = (int) (val % totalWeight);

        // Binary search cumulative weights
        int low = 0;
        int high = count - 1;
        while (low < high) {
            int mid = (low + high) >>> 1;
            long slotBase = probableObjectsPtr + 8L + (mid * 16L);
            int cumWeight = ForeignMemory.getInt(slotBase + 8L);
            if (cumWeight < r) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        long slotBase = probableObjectsPtr + 8L + (low * 16L);
        return ForeignMemory.getLong(slotBase);
    }

    // ==========================================
    // PARAMETERLESS SYSTEM_RNG HELPER OVERLOADS
    // ==========================================

    public static long nextLong() {
        return nextLong(SYSTEM_RNG);
    }

    public static int nextInt() {
        return nextInt(SYSTEM_RNG);
    }

    public static float nextFloat() {
        return nextFloat(SYSTEM_RNG);
    }

    public static double nextDouble() {
        return nextDouble(SYSTEM_RNG);
    }

    public static float nextNDCFloat() {
        return nextNDCFloat(SYSTEM_RNG);
    }

    public static char nextChar() {
        return nextChar(SYSTEM_RNG);
    }

    public static boolean getWeight(int weight, int total) {
        return getWeight(SYSTEM_RNG, weight, total);
    }

    public static long sample(long probablePtr) {
        return sample(SYSTEM_RNG, probablePtr);
    }

    public static long probablePool(long probableObjectsPtr) {
        return probablePool(SYSTEM_RNG, probableObjectsPtr);
    }
}

File: src/util/random.c
#include "util/random.h"

#include <stdlib.h>

#include <mach/mach_time.h>

#include "nio/mem.h"
#include "oop/type.h"
#include "util/hash.h"
#include "annotation/overview.h"

;;OVERVIEW
/**
 * ============================================================================
 * MODULE: Random (util/random.c)
 * LEVEL: L2 — Behavior (utility behavior API)
 * ============================================================================
 * the Random class, ported from util/Random.java.
 *
 * FUNCTION REGISTRY:
 * ----------------------------------------------------------------------------
 * Constructors:
 *   - Random_1(seed)
 *
 * Core Functions:
 *   - Random_free(r)
 *   - Random_system(void)
 *   - Random_nextLong(r)
 *   - Random_nextInt(r)
 *   - Random_nextFloat(r)
 *   - Random_nextDouble(r)
 *   - Random_nextNDCFloat(r)
 *   - Random_nextChar(r)
 *   - Random_sample(r, probable)
 *   - Random_probablePool(r, pool)
 *
 * Getters:
 *   - Random_getWeight(r, weight, total)
 * ============================================================================
 */


// random.c — Random port (Legacy: util/Random.java).

static const uint64_t GOLDEN_RATIO_64 = 0x9e3779b97f4a7c15ull;

static Random *system_rng = nullptr;

static uint64_t system_seed(void) {
    uint64_t ticks = (uint64_t)mach_absolute_time();
    return ticks ^ (uint64_t)(uintptr_t)&system_rng;
}

Random *Random_1(uint64_t seed) {
    Random *r = (Random*) Memory_alloc(TYPE_RANDOM, sizeof(Random));
    if (!r) return nullptr;
    (*r).seed = seed;
    (*r).counter = 0;
    (*r).pad = 0;
    return r;
}

void Random_free(Random *r) {
    Memory_free(r);
}

Random *Random_system(void) {
    if (!system_rng)
        system_rng = Random_1(system_seed());
    return system_rng;
}

uint64_t Random_nextLong(Random *r) {
    if (!r) return 0;
    uint64_t mixed = (*r).seed
        ^ (uint64_t)(uintptr_t)r
        ^ ((uint64_t)(*r).counter * GOLDEN_RATIO_64);
    uint64_t result = Hash_murmur3Mix64(mixed);
    (*r).seed = result;
    (*r).counter++;
    return result;
}

int32_t Random_nextInt(Random *r) {
    return (int32_t)Random_nextLong(r);
}

float Random_nextFloat(Random *r) {
    return (float)((Random_nextLong(r) & 0xFFFFFF) / 16777216.0);
}

double Random_nextDouble(Random *r) {
    return (double)((Random_nextLong(r) & 0x1FFFFFFFFFFFFFull) / 9007199254740992.0);
}

float Random_nextNDCFloat(Random *r) {
    return (float)(((Random_nextLong(r) & 0xFFFFFF) / 8388608.0) - 1.0);
}

char Random_nextChar(Random *r) {
    int64_t v = (int64_t)Random_nextLong(r);
    return (char)(32 + (llabs(v) % 95));
}

bool Random_getWeight(Random *r, uint32_t weight, uint32_t total) {
    if (!r) return false;
    if (total == 0) return false;
    if (weight >= total) return true;
    if (weight == 0) return false;
    uint64_t val = Random_nextLong(r) & 0x7FFFFFFFFFFFFFFFull;
    return (val % total) < weight;
}

uintptr_t Random_sample(Random *r, const Probable *probable) {
    if (!probable) return 0;
    if (Random_getWeight(r, (*probable).weight, (*probable).total))
        return (*probable).object;
    return 0;
}

uintptr_t Random_probablePool(Random *r, const ProbableObjects *pool) {
    if (!pool) return 0;
    size_t count = ProbableObjects_size((ProbableObjects*) pool);
    if (count == 0) return 0;

    uint32_t total_weight = ProbableObjects_totalWeight((ProbableObjects*) pool);
    uint64_t val = Random_nextLong(r) & 0x7FFFFFFFFFFFFFFFull;

    if (total_weight == 0) {
        size_t idx = (size_t)(val % count);
        return ProbableObjects_objectAt((ProbableObjects*) pool, idx);
    }

    uint32_t target = (uint32_t)(val % total_weight);
    size_t low = 0;
    size_t high = count - 1;
    while (low < high) {
        size_t mid = (low + high) / 2;
        uint32_t cumulative = ProbableObjects_cumulativeAt((ProbableObjects*) pool, mid);
        if (cumulative < target)
            low = mid + 1;
        else
            high = mid;
    }
    return ProbableObjects_objectAt((ProbableObjects*) pool, low);
}
